Fixed crash of the LDAP authorization process if there are parentheses in the user_dn variable

IlyaTsoi opened the pull request:
Create a function escapeForLDAPFilter and use it to escape characters '(' and ')' in a final_user_dn variable that we receive from the LDAP server. This is necessary if these characters are present in the user's DN. Otherwise, the entire LDAP authorization process crashes with the error "Bad serach filter"

Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Create a function escapeForLDAPFilter and use it to escape characters '(' and ')' in a final_user_dn variable

Detailed description / Documentation draft:
Create a function escapeForLDAPFilter and use it to escape characters '(' and ')' in a final_user_dn variable that we receive from the LDAP server. This is necessary if these characters are present in the user's DN. Otherwise, the entire LDAP authorization process crashes with the error "Bad serach filter". The error was found when setting up authorization in the Clickhouse through LDAP Active Directory server


> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

> If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.


> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

IlyaTsoi committed:
Update LDAPClient.cpp

Create a function escapeForLDAPFilter and use it to escape characters '(' and ')' in a variable that we receive from the LDAP server. This is necessary if these characters are present in the user's DN. Otherwise, the entire LDAP authorization process crashes with the error "Bad serach filter"

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=33401) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=33401) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=33401) it.</sub>

IlyaTsoi committed:
Update LDAPClient.cpp

refactoring of escapeForLDAPFilter function

IlyaTsoi committed:
The most minimalistic solution

You can simply add parentheses to the switch statement to escapeForLDAP function and apply it when assigning a value to the final_user_dn variable. This fixes the error, I checked

IlyaTsoi committed:
Insert missing word

Insert missing word 'секцию'

IlyaTsoi committed:
Revert "The most minimalistic solution"

This reverts commit 9472669dd59b255f1adfc997860113ccc3969637.

IlyaTsoi committed:
Revert "Update LDAPClient.cpp"

This reverts commit 30ef16ab58709ef9522b3d7eb62d74513cc6b4ca.

vitlibar commented:
@traceon Please take a look

traceon reviewed (changes requested):
Left inline comments.

traceon commented on src/Access/LDAPClient.cpp:
```diff
@@ -319,7 +343,7 @@ void LDAPClient::openConnection()
                     throw Exception("Failed to detect user DN: more than one entry in the search results", ErrorCodes::LDAP_ERROR);
 
-                final_user_dn = *user_dn_search_results.begin();
+                final_user_dn = escapeForLDAPFilter(*user_dn_search_results.begin());
```
Also, I would suggest moving `escapeForFilter` calls to:
```c++
    const auto final_search_filter = replacePlaceholders(search_params.search_filter, {
        {"{user_name}", escapeForFilter(final_user_name)},
        {"{bind_dn}", escapeForFilter(final_bind_dn)},
        {"{user_dn}", escapeForFilter(final_user_dn)},
        {"{base_dn}", escapeForFilter(final_base_dn)}
    });
```
instead doing it here when assigning.

IlyaTsoi replied:
Yes, you are right )

traceon commented on src/Access/LDAPClient.cpp:
```diff
@@ -95,6 +95,30 @@ namespace
 
         return dest;
     }
+    
```
Remove trailing spaces:
```suggestion

```

IlyaTsoi commented on src/Access/LDAPClient.cpp:
```diff
@@ -95,6 +95,30 @@ namespace
+        {
+            switch (ch)
+            {
+                case '(':
+                    dest += "\\28";
+                    break;
+                case ')':
+                    dest += "\\29";
+                    break;
+                default:
+                    dest += ch;
+                    break;
```
Yes, it's right

IlyaTsoi committed:
Update src/Access/LDAPClient.cpp

escape all special chars

Co-authored-by: Denis Glazachev <traceon@users.noreply.github.com>

IlyaTsoi commented on src/Access/LDAPClient.cpp:
```diff
@@ -95,6 +95,30 @@ namespace
         return dest;
     }
+    
+    auto escapeForLDAPFilter(const String & src)
```
OK

IlyaTsoi committed:
Update src/Access/LDAPClient.cpp

rename function

Co-authored-by: Denis Glazachev <traceon@users.noreply.github.com>

IlyaTsoi committed:
Update src/Access/LDAPClient.cpp

Remove trailing spaces:

Co-authored-by: Denis Glazachev <traceon@users.noreply.github.com>

traceon commented on src/Access/LDAPClient.cpp:
```diff
@@ -95,8 +95,7 @@ namespace
     }
-    
-    auto escapeForLDAPFilter(const String & src)
+    auto escapeForFilter(const String & src)
```
```suggestion

    auto escapeForFilter(const String & src)
```

IlyaTsoi committed:
Update src/Access/LDAPClient.cpp

rename function

Co-authored-by: Denis Glazachev <traceon@users.noreply.github.com>

IlyaTsoi committed:
Update src/Access/LDAPClient.cpp

Co-authored-by: Denis Glazachev <traceon@users.noreply.github.com>

IlyaTsoi committed:
Update LDAPClient.cpp

use escapeForFilter when assigning value to const final_search_filter

traceon reviewed (approved):
LGTM, thanks. You can do this too, if you fancy: https://github.com/ClickHouse/ClickHouse/pull/33401#discussion_r784077727

IlyaTsoi committed:
Update LDAPClient.cpp

rename function escapeForLDAP to escapeForDN

IlyaTsoi commented:
> LGTM, thanks. You can do this too, if you fancy: [#33401 (comment)](https://github.com/ClickHouse/ClickHouse/pull/33401#discussion_r784077727)

OK, i don't mind )

vitlibar commented on src/Access/LDAPClient.cpp:
```diff
@@ -96,6 +96,39 @@ namespace
+    auto escapeForFilter(const String & src)
+    {
+        String dest;
+        dest.reserve(src.size() * 2);
```
It seems we replace some characters with three characters in the code below, so I think we should multiply `src.size()` by `3` here, not by `2`.

IlyaTsoi replied:
Yes, you a right ), fixed

IlyaTsoi committed:
Update LDAPClient.cpp

Change volume of reserved memory for variable

vitlibar commented:
The failed tests are not related to this PR's changes.

traceon commented:
@vzakaznikov could you please look into adding a simple case to cover this?

vzakaznikov commented:
Yes, I will add a test for it.
